node/src/blockchain/blockchain_interface/blockchain_interface_web3/lower_level_interface_web3.rs

```diff
@@ -1,395 +1,762 @@
+        )
+    }
+
+    fn get_transaction_receipt(
+        &self,
+        hash: H256,
+    ) -> Box<dyn Future<Item = Option<TransactionReceipt>, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .transaction_receipt(hash)
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
+    }
+
+    fn get_transaction_receipt_batch(
+        &self,
+        hash_vec: Vec<H256>,
+    ) -> Box<dyn Future<Item = Vec<TransactionReceiptResult>, Error = BlockchainError>> {
+        let _ = hash_vec.into_iter().map(|hash| {
+            self.web3_batch.eth().transaction_receipt(hash);
+        });
+        Box::new(
+            self.web3_batch
+                .transport()
+                .submit_batch()
+                .map_err(|e| QueryFailed(e.to_string()))
+                .and_then(move |batch_response| {
+                    Ok(batch_response
+                        .into_iter()
+                        .map(|response| match response {
+                            Ok(result) => {
+                                match serde_json::from_value::<TransactionReceipt>(result) {
+                                    Ok(receipt) => TransactionReceiptResult::Found(receipt),
+                                    Err(e) => {
+                                        if e.to_string().contains("invalid type: null") {
+                                            TransactionReceiptResult::NotPresent
+                                        } else {
+                                            TransactionReceiptResult::Error(e.to_string())
+                                        }
+                                    }
+                                }
+                            }
+                            Err(e) => TransactionReceiptResult::Error(e.to_string()),
+                        })
+                        .collect::<Vec<TransactionReceiptResult>>())
+                }),
+        )
+    }
+
+    fn get_contract(&self) -> Contract<Http> {
+        self.contract.clone()
+    }
+
+    fn get_transaction_logs(
+        &self,
+        filter: Filter,
+    ) -> Box<dyn Future<Item = Vec<Log>, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .logs(filter)
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
+    }
+
+    fn submit_payables_in_batch(
+        &self,
+        logger: Logger,
+        chain: Chain,
+        consuming_wallet: Wallet,
+        fingerprints_recipient: Recipient<PendingPayableFingerprintSeeds>,
+        affordable_accounts: Vec<PayableAccount>,
+    ) -> Box<dyn Future<Item = Vec<ProcessedPayableFallible>, Error = PayableTransactionError>>
+    {
+        let web3_batch = self.web3_batch.clone();
+        let get_transaction_id = self.get_transaction_id(consuming_wallet.address());
+        let get_gas_price = self.get_gas_price();
```

Comment:
Can you explain why you need to query the already known transaction id and gas_price again. It's both stored inside the blockchain agent. 
If the payment adjuster confirms payments coverage by our balances, using quite precise computation, why would you change the inputs by updating the gas price. For Utkarsh: we agreed that any possible momentary fluctuation is going to be handled thanks to the added margin.

This second call of get_gas_price is unnecessary. Please remove it. 

The call of get_transaction_id is probably reasonable put as close as possible to the place where the payments are taking place. I'm happy with that. What I'm not, though, it's the blockchain agent constructor because it also includes this operation. Please make sure we do that only once, I don't want to see it repeat because it's so obvious we need to do it only once.

Reply:
**URGENCY STATUS: THIS IS NOT GOOD, BUT A BIG MISTAKE, THAT HAPPENED IN THIS CARD SO IT SHOULD BE FIXED HERE TOO** You know, let's allow adding in just useful new stuff, not together with also new issues.

Reply:
We have added the explanation to this as a comment in the code.

Reply:
I'm afraid, you may've misunderstood to me. I can't imagine why this could be solved by a comment. The double calls are wrong, both. 

If you want to do the second call of `get_gas_price()` you'd have to change the whole concept.
Maybe something like 

[<Creating blockchain agent>;  first gas price fetch] ---> 

[<PaymentAdjusterOperations>; fetched gas price augmented by tx_fee_margin (15% at the moment); PA checks for insolvency, the augmented gas price is stored in the OutboundPaymentInstructions msg ] ---> 

[<Before forming the batch of transactions>; You fetch gas price the second time. You don't proceed straight, you must first check the following condition **gas_price1 + margin >= gas_price2**, if it's correct you can use gas_price2, if it's false, you have to throw out an error. Why? Because we are already behind the checks of PA and if you increase the gas_price it threats with transaction failures which PA is meant to prevent ]  

So, you either implement this up here :point_up_2: or you need to remove this second call completely. There's no place for it after we leave the PaymentAdjuster behind. In the simplest solution (not like the more advanced depicted above) we always fetch the gas price only before the PaymentAdjuster.

Reply:
We plan on removing this second call completely once `PaymentAdjuster` is in.
we did it in this way because we were fetching the value from the database previously since we aren't storing the values anywhere we need to fetch it from the RPC.
Now this solutions is a mea filling the blank so that we don't break any code.
Ideally blockchain agent will provide the value.

feel free to modify this under your card that fits the new design of `PaymentAdjuster`

Reply:
I'm a bid frustrated that you have decided so. Last time I talked to Utkarsh, it was on the meeting with Dan (Syther is excused), he seemed quite assured when I showed him that this call must (there aren't any doubts) while the BlockchainAgent is being made.  It definitely needs to disappear here. Utkarsh even wrote down some TODOs in his IDE to remind him this. What I told him too though that I don't care so much about location of `get_transaction_id()`. If here or together with the get_gas_price() earlier, it's not a big difference. 

You say "ideally blockchain agent will provide the value", but it has ever been designed like so. That's the problem, you're breaking the already established order of things. The BlockchainAgent was deployed beforehand, not requiring the PaymentAdjuster to be functioning. 
Now, out of a sudden, with your card, the operations are doubled. So why should anybody else but you two fix this to be right? 

I urge you to take an excursion to the master branch and check these functions as they go in sequence: 
node/src/blockchain/blockchain_bridge.rs **handle_outbound_payments_instructions**         
node/src/blockchain/blockchain_bridge.rs **process_payments**
node/src/blockchain/blockchain_interface/blockchain_interface_web3/mod.rs **send_batch_of_payables**
Notice how the agent is being used. Also watch where there are calls done by the `BlockchainInterface`. The answer is "only once ever". Compare it with your three calls. Also notice how the method send_batch_of_payables cooperates with the supplied agent. It asks, and the agent has answers. There is no interaction with the Internet. The agent knows all that is needed. This wasn't accidentally designed. It is well-thought and has good reasons. 
In your code, you go about using some recklessly written code. This is as if going backward, do you know what I mean.

Remember that this is why the agent exists. It keeps knowledge of facts coming from the chain currently being used. The agent works better as immutable. Thus, the idea was to collect important information given the chain and then it was stuffed into the agent. The agent travels by messages and anywhere where somebody is interested in some blockchain specific things the agent is on hand. It's the whole point of its existence. 
But you don't follow, you disregard the original concept. :man_shrugging: Simply get rid of the second-place RPC calls and let the agent serve its purpose...  

So at least don't tell me that you want to preserve some existing code, because these are all your new additions and I'm rightful complaining if you introduce function calls which actually double some work already done.
It's silly that I even have to carry these long discussions. After the months of work you, should have understood the design you took over at the starting line.

(If there are now any tests that you - perhaps - wrote for these calls above the misconception, of course, delete them, as they, together with the doubled calls, don't belong in the codebase)

If you feel it's over your head, ask help from Dan and his supervision.

Reply:
We have removed the get_transaction_id call from blockchain agent.
and we have also removed the get_gas_price from submit batch.